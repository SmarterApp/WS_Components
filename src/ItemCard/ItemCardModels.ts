import * as GradeLevels from '../GradeLevels/GradeLevels';
import { Link } from 'react-router-dom';

export function itemPageLink(bankKey: number, itemKey: number) {
    window.location.href = "/Item?bankKey=" + bankKey + "&itemKey=" + itemKey;
}

export interface ItemCardModel {
    bankKey: number;
    itemKey: number;
    title: string;
    grade: GradeLevels.GradeLevels;
    gradeLabel: string;
    subjectCode: string;
    subjectLabel: string;
    claimCode: string;
    claimLabel: string;
    targetHash: number;
    targetShortName: string;
    interactionTypeCode: string;
    interactionTypeLabel: string;
    isPerformanceItem: boolean;
}
