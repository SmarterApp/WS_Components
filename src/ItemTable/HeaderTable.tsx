import * as React from "react";
import {
  SortColumnModel,
  HeaderSortModel,
  SortDirection
} from "./ItemTableModels";

/**
 * Properties for HeaderTable component
 * @interface HeaderTableProps
 */
export interface HeaderTableProps {
  columns: SortColumnModel[];
  onHeaderClick: (header: SortColumnModel) => void;
  sorts: HeaderSortModel[];
}

const invokeResetSortLimit = 2;

const style = {
  color: "gray"
};

const descendingArrow = (
  <span style={style} className="fa fa-sort-desc" aria-hidden="true" />
);

const ascendingArrow = (
  <span style={style} className="fa fa-sort-asc" aria-hidden="true" />
);

const noSort = <span style={style} className="fa fa-sort" aria-hidden="true" />;

/**
 * HeaderTable creates a table header based on the passed in columns
 * The HeaderTable, when clicked, will add the clicked column header
 * the parent, ItemTableContainer, state, sorting the table
 */
export class HeaderTable extends React.Component<HeaderTableProps, {}> {
  constructor(props: HeaderTableProps) {
    super(props);
    this.state = {
      dirElem: undefined
    };
  }
  /**
   * Send the clicked sort column to the parent to be added to the sorts list
   * using onHeaderClick prop.
   * @param {SortColumnModel} sCol
   * @param {(HeaderSortModel | undefined)} hCol
   */
  headerEventHandler(sCol: SortColumnModel, hCol: HeaderSortModel | undefined) {
    this.props.onHeaderClick(sCol);
  }

  /**
   * Assigns an ascending or descending arrow character to the visible column
   * header when clicked, denoting how it is sorted
   * @param {(HeaderSortModel | undefined)} headerSort
   * @returns {JSX.Element}
   */
  setDirElem(headerSort: HeaderSortModel | undefined): JSX.Element {
    let dirElem = noSort;
    if (!headerSort) {
      return dirElem;
    } else if (headerSort.direction === SortDirection.Ascending) {
      dirElem = ascendingArrow;
    } else if (headerSort.direction === SortDirection.Descending) {
      dirElem = descendingArrow;
    }
    return dirElem;
  }

  /**
   * Renders a single table header element and the corresponding ascending, descending,
   * or 'not-sorted' symbol
   * @param {SortColumnModel} col
   * @returns {JSX.Element}
   */
  renderHeader(col: SortColumnModel): JSX.Element {
    const headerSort = this.props.sorts.find(
      hs => hs.col.header === col.header
    );
    if (headerSort) {
      this.setDirElem(headerSort);
    }

    return (
      <th
        key={col.header}
        className={col.className}
        onClick={() => this.headerEventHandler(col, headerSort)}
      >
        <div className={col.className}>
          <span>{col.header}</span>
          {this.setDirElem(headerSort)}
        </div>
      </th>
    );
  }

  render() {
    return (
      <thead>
        <tr className="primary">
          <th />
          {this.props.columns.map(col => this.renderHeader(col))}
        </tr>
      </thead>
    );
  }
}
