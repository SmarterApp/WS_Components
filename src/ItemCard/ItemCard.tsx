import * as React from "react";
import * as GradeLevels from "../GradeLevels/GradeLevels";
import * as ItemCardModels from "./ItemCardModels";
import { Redirect } from "react-router";

export interface ItemCardState {
  redirect: boolean;
}
export interface ItemCardProps extends ItemCardModels.ItemCardModel {}

export class ItemCard extends React.Component<ItemCardProps, ItemCardState> {
  constructor(props: ItemCardProps) {
    super(props);
    this.state = { redirect: false };
  }

  shouldComponentUpdate(nextProps: ItemCardProps, nextState: ItemCardState) {
    return nextState.redirect;
  }

  handleKeyPress = (e: React.KeyboardEvent<HTMLElement>) => {
    if (e.keyCode === 13 || e.keyCode === 23) {
      this.setState({ redirect: true });
    }
  };

  handleOnClick = () => {
    this.setState({ redirect: true });
  };

  render() {
    const { bankKey, itemKey } = this.props;
    let content = <Redirect push to={`/Item/${bankKey}-${itemKey}`} />;
    if (!this.state.redirect) {
      content = (
        <div
          role="link"
          className={`card card-block ${this.props.subjectCode.toLowerCase()}`}
          onClick={this.handleOnClick}
          onKeyUp={this.handleKeyPress}
          tabIndex={0}
        >
          <div className="card-contents">
            <h4 className="card-title">{this.props.title}</h4>
            <p className="card-text subject">
              <span className="card-text-label">Subject:</span>
              <span className="card-text-value">
                {" "}
                {this.props.subjectLabel}
              </span>
            </p>
            <p className="card-text grade">
              <span className="card-text-label">Grade:</span>
              <span className="card-text-value"> {this.props.gradeLabel}</span>
            </p>
            <p className="card-text claim">
              <span className="card-text-label">Claim:</span>
              <span className="card-text-value"> {this.props.claimLabel}</span>
            </p>
            <p className="card-text target">
              <span className="card-text-label">Target:</span>
              <span className="card-text-value">
                {this.props.targetShortName}
              </span>
            </p>
            <p className="card-text interaction-type">
              <span className="card-text-label">Item Type:</span>
              <span className="card-text-value">
                {` ${this.props.interactionTypeLabel}`}
              </span>
            </p>
            <p className="card-text item-id">
              <span className="card-text-label">Item Id:</span>
              <span className="card-text-value"> {this.props.itemKey}</span>
            </p>
          </div>
        </div>
      );
    }

    return content;
  }
}
