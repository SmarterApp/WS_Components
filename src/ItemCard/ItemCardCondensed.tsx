import "src/Assets/Styles/item-card.less";
import * as React from "react";
import * as ItemCardModels from "./ItemCardModels";
import { Redirect } from "react-router";

export interface ItemCardCondensedState {
  redirect: boolean;
}
export interface ItemCardCondensedProps extends ItemCardModels.ItemCardModel {}

export class ItemCardCondensed extends React.Component<
  ItemCardCondensedProps,
  ItemCardCondensedState
> {
  constructor(props: ItemCardCondensedProps) {
    super(props);
    this.state = { redirect: false };
  }

  shouldComponentUpdate(
    nextProps: Readonly<ItemCardCondensedProps>,
    nextState: Readonly<ItemCardCondensedState>,
    nextContext: any
  ): boolean {
    return this.state.redirect;
  }

  handleKeyPress(e: React.KeyboardEvent<HTMLElement>) {
    if (e.keyCode === 13 || e.keyCode === 23) {
      this.setState({ redirect: true });
    }
  }
  handleOnClick() {
    this.setState({ redirect: true });
  }

  render() {
    const { bankKey, itemKey } = this.props;
    if (this.state.redirect) {
      return <Redirect push to={`/Item/${bankKey}-${itemKey}`} />;
    }
    return (
      <div
        className={`card card-block ${this.props.subjectCode.toLowerCase()} condensed`}
        onClick={() => this.handleOnClick()}
        onKeyUp={e => this.handleKeyPress(e)}
        tabIndex={0}
      >
        <div className="card-contents">
          <h4 className="card-title">{this.props.subjectLabel}</h4>
          <p className="card-text claim">
            <span className="card-text-label">Claim:</span>
            <span className="card-text-value"> {this.props.claimLabel}</span>
          </p>
          <p className="card-text target">
            <span className="card-text-label">Target:</span>
            <span className="card-text-value">
              {" "}
              {this.props.targetShortName}
            </span>
          </p>
          <p className="card-text interaction-type">
            <span className="card-text-label">Item Type:</span>
            <span className="card-text-value">
              {" "}
              {this.props.interactionTypeLabel}
            </span>
          </p>
          <p className="card-text item-id">
            <span className="card-text-label">Item Id:</span>
            <span className="card-text-value"> {this.props.itemKey}</span>
          </p>
        </div>
      </div>
    );
  }
}
