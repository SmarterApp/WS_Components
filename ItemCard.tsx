import * as React from 'react';
import * as GradeLevels from './GradeLevels';
import * as API from './ApiModels';
import * as AboutItem from './AboutItem';
import * as ItemCardViewModel from './ItemCardViewModel'
import * as ItemCardFields from './ItemCardFields'
import * as ItemCardCondensed from './ItemCardCondensed'

export interface Props{
    aboutItem: ItemCardViewModel.ItemCardViewModel;
}

export class ItemCard extends React.Component<Props, {}> {
    handleKeyPress(bankKey: number, itemKey: number, e: React.KeyboardEvent<HTMLElement>) {
        if (e.keyCode === 13 || e.keyCode === 23) {
            ItemCardFields.itemPageLink(bankKey, itemKey);
        }
    }

    shouldComponentUpdate(nextProps: Readonly<Props>, nextState: Readonly<{}>, nextContext: any): boolean {
        return false;
    }

    renderCard(bankKey: number, itemKey: number) {
        return (
            <div className={`card card-block ${this.props.aboutItem.subjectCode.toLowerCase()}`}
                onClick={e => ItemCardFields.itemPageLink(bankKey, itemKey)}
                onKeyUp={e => this.handleKeyPress(bankKey, itemKey, e)}
                tabIndex={0}>
                <div className="card-contents">
                    <ItemCardFields.ItemCardFields aboutItem={this.props.aboutItem}/>
                </div>
            </div>
        );
    }

    render() {
        return (
            <div>
                {this.renderCard(this.props.aboutItem.bankKey, this.props.aboutItem.itemKey)};
            </div>
        );
    }
}