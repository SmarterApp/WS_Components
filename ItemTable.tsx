import * as ItemCard from "./ItemCard";
import * as ItemCardViewModel from './ItemCardViewModel';
import * as React from "react";

export type Header = "Item" | "Claim/Target" | "Subject" | "Grade" | "Item Type";

export enum SortDirection {
    NoSort = 0,
    Ascending = 1,
    Descending = -1
}

export interface HeaderSort {
    col: SortColumn;
    direction: SortDirection;
    resetSortCount: number;
}

export interface SortColumn {
    header: Header;
    className: string;
    accessor: (label: ItemCardViewModel.ItemCardViewModel) => string | number;
    compare: (a: ItemCardViewModel.ItemCardViewModel, b: ItemCardViewModel.ItemCardViewModel) => number;
}

const invokeResetSortLimit = 2;

const decendingArrow = (
    <span aria-label="Decend">▼</span>
);

const acendingArrow = (
    <span aria-label="Ascend">▲</span>
);

const noSort = (
    <span aria-label="NoSort"></span>
);

export const headerColumns: SortColumn[] = [
    {
        header: "Item",
        className: "item",
        accessor: label => label.itemKey,
        compare: (a, b) => (a.itemKey) - (b.itemKey)
    },
    {
        header: "Claim/Target",
        className: "claimAndTarget",
        accessor: label => label.claimLabel + "/" + label.target,
        compare: (a, b) => {
            if (a.claimCode < b.claimCode || a.target < b.target) {
                return 1
            }
            else if (a.claimCode > b.claimCode || a.target > b.target) {
                return -1
            }
            else {
                return 0;
            }
        }
    },
    {
        header: "Subject",
        className: "subject",
        accessor: label => label.subjectLabel,
        compare: (a, b) => (a.subjectCode).localeCompare(b.subjectCode)
    },
    {
        header: "Grade",
        className: "grade",
        accessor: label => label.gradeLabel,
        compare: (a, b) => (a.grade) - (b.grade)
    },
    {
        header: "Item Type",
        className: "interactionType",
        accessor: label => label.interactionTypeLabel,
        compare: (a, b) => (a.interactionTypeCode).localeCompare(b.interactionTypeCode)
    },
];

interface HeaderProps {
    columns: SortColumn[];
    onHeaderClick: (header: SortColumn) => void;
    sorts: HeaderSort[];
}

export class HeaderTable extends React.Component<HeaderProps, {}> {
    constructor(props: HeaderProps){
        super(props);
    }

    compareColumn(lhs: ItemCardViewModel.ItemCardViewModel, rhs: ItemCardViewModel.ItemCardViewModel): number {
        const sorts = this.props.sorts || [];
        for (const sort of sorts) {
            const diff = sort.col.compare(lhs, rhs) * sort.direction;
            if (diff !== 0) {
                return diff;
            }
        }
        return 0;
    }

    headerEventHandler(scol: SortColumn, hcol: HeaderSort|undefined) {
        this.props.onHeaderClick(scol);
    }

    renderHeader(col: SortColumn): JSX.Element {
        let dirElem: string | undefined | JSX.Element;
        const headerSort = this.props.sorts.find(hs => hs.col.header === col.header);
        if (!headerSort) {
            dirElem = noSort;
        } else if (headerSort.direction === SortDirection.Ascending) {
            dirElem = acendingArrow;
        } else if (headerSort.direction === SortDirection.Descending) {
            dirElem = decendingArrow;
        } else {
            dirElem = noSort;
        }

        return (
            <th key={col.header}
                className={col.className}
                onClick={() => this.headerEventHandler(col,headerSort)}>
                <div className={col.className}>
                    {dirElem} {col.header}
                </div>
            </th>
        );
    }

    render() {
        return (
            <table className="item-table table mapcomponent-table">
                <thead>
                    <tr className="primary">
                        <th></th>
                        {this.props.columns.map(col => this.renderHeader(col))}
                    </tr>
                </thead>
            </table>
        );
    }
}

interface Props {
    tableRef?: (ref: HTMLTableElement) => void;
    mapRows: ItemCardViewModel.ItemCardViewModel[];
    rowOnClick: (item: ItemCardViewModel.ItemCardViewModel) => void;
    sort: HeaderSort[];
    columns: SortColumn[];
    selectedRow?: ItemCardViewModel.ItemCardViewModel;
}

export class DataTable extends React.Component<Props, {}> {
    constructor(props: Props) {
        super(props);
    }

   

    renderCell(col: SortColumn, cellData: ItemCardViewModel.ItemCardViewModel): JSX.Element {
        return (
            <td key={col.header}
                className={col.className}>
                <div className={col.className}>
                    {col.accessor(cellData)}
                </div>
            </td>
        );
    }

    //TODO replace X with checkmark icon 
    renderRow(rowData: ItemCardViewModel.ItemCardViewModel, index: number): JSX.Element {
        let isSelected = false;
        if (this.props.selectedRow) {
            isSelected = rowData.itemKey === this.props.selectedRow.itemKey
                        && rowData.bankKey === this.props.selectedRow.bankKey;
        }
        
        return (
            <tr key={index} className={isSelected ? "selected" : ""}
                onClick={() => {
                    this.props.rowOnClick(rowData);
                }}>
                <td>{isSelected ? "X" : ""}</td>
                {this.props.columns.map(col => this.renderCell(col, rowData))}
            </tr>
        );
    }

    renderRows(): JSX.Element {
        const rows = this.props.mapRows.map((rowData,idx) => this.renderRow(rowData,idx));
        return (
            <tbody>{rows}</tbody>
        );
    }

    render() {
        return (
            <table className="item-table table table-striped mapcomponent-table"
                ref={this.props.tableRef}>
                {this.renderRows()}
            </table>
        );
    }
}