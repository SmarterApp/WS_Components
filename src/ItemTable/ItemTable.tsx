import * as React from "react";
import { HeaderSortModel, SortColumnModel } from "./ItemTableModels";
import { Resource } from "../ApiModel";
import { AboutItemModel } from "../AboutItem/AboutItemModels";
import { ItemCardModel } from "../ItemCard/ItemCardModels";
import { ItemCardViewer } from "../ItemCard/ItemCardViewer";
import * as ReactTooltip from "react-tooltip";
import { findDOMNode } from "react-dom";

export interface ItemTableProps {
  mapRows: ItemCardModel[];
  onRowExpand: (item: ItemCardModel) => void;

  onRowSelect: (item: ItemCardModel) => void;
  sort: HeaderSortModel[];
  columns: SortColumnModel[];
  expandedRow?: ItemCardModel;
  item?: Resource<AboutItemModel>;

  isLinkTable: boolean;
}
/**
 * Renders the table populated from an array of ItemCardModels. Also renders an instance of the ItemCardViewer,
 * inserting a responsive sub-table with an iframe that displays the Item Card.
 * @class ItemTable
 * @extends {React.Component<ItemTableProps, {}>}
 */
export class ItemTable extends React.Component<ItemTableProps, {}> {
  constructor(props: ItemTableProps) {
    super(props);
  }

  collapse = (
    <i className="fa fa-chevron-right fa-sm table-icon" aria-hidden="true" />
  );
  expand = (
    <i className="fa fa-chevron-down fa-sm table-icon" aria-hidden="true" />
  );

  handleRowClick = (rowData: ItemCardModel) => {
    this.props.onRowExpand(rowData);
  };

  handleKeyUpEnter = (e: React.KeyboardEvent<any>, rowData: ItemCardModel) => {
    if (e.keyCode === 13) {
      this.props.onRowExpand(rowData);
    }
  };

  handleCheckboxClick = (
    e: React.MouseEvent<HTMLTableDataCellElement>,
    rowData: ItemCardModel
  ) => {
    e.stopPropagation();
    this.props.onRowSelect(rowData);
  };

  handleCheckboxKeyUpEnter = (
    e: React.KeyboardEvent<any>,
    rowData: ItemCardModel
  ) => {
    if (e.keyCode === 13) {
      e.stopPropagation();
      this.props.onRowSelect(rowData);
    }
  };

  handleToolTipKeyUpEnter = (e: React.KeyboardEvent<any>) => {
    if (e.keyCode === 13) {
      e.stopPropagation();
      ReactTooltip.show(findDOMNode(e.currentTarget));
    }
  };

  renderCell(col: SortColumnModel, cellData: ItemCardModel): JSX.Element {
    let tag: JSX.Element;

    // TODO: implement description API functionality when available.
    const descriptionAvailable = true;
    if (descriptionAvailable) {
      tag = (
        <div className={`${col.className}-${cellData.itemKey}`}>
          <a
            tabIndex={0}
            onClick={e => e.stopPropagation()}
            onKeyUp={e => this.handleToolTipKeyUpEnter(e)}
            data-for={`tooltip-${col.className}-${cellData.itemKey}`}
            data-event="click"
            data-tip="custom tooltip here"
            role="button"
          >
            {col.accessor(cellData)}
          </a>
          <ReactTooltip
            id={`tooltip-${col.className}-${cellData.itemKey}`}
            globalEventOff="focusout"
          />
        </div>
      );
    } else {
      tag = <div className={col.className}>{col.accessor(cellData)}</div>;
    }

    return (
      <td key={col.header} className={col.className}>
        {tag}
      </td>
    );
  }

  renderRow(rowData: ItemCardModel, index: number): JSX.Element[] {
    const { expandedRow, columns, item, isLinkTable } = this.props;
    const unChecked = (
      <i className="fa fa-square-o fa-sm table-icon" aria-hidden="true" />
    );
    const checked = (
      <i className="fa fa-check-square-o fa-sm table-icon" aria-hidden="true" />
    );
    let isExpanded = false;
    if (expandedRow) {
      isExpanded =
        rowData.itemKey === expandedRow.itemKey &&
        rowData.bankKey === expandedRow.bankKey;
    }

    const controls: JSX.Element[] = [];
    if (!isLinkTable) {
      controls.push(
        <td
          key={rowData.bankKey}
          onClick={e => this.handleCheckboxClick(e, rowData)}
          onKeyUp={e => this.handleCheckboxKeyUpEnter(e, rowData)}
          tabIndex={0}
        >
          {rowData.selected === true ? checked : unChecked}&nbsp;
        </td>,
        <td key={rowData.itemKey}>
          {isExpanded ? this.expand : this.collapse}
        </td>
      );
    }
    const row: JSX.Element[] = [
      <tr
        key={index}
        className={isExpanded ? "selected" : ""}
        onClick={() => this.handleRowClick(rowData)}
        onKeyUp={e => this.handleKeyUpEnter(e, rowData)}
        tabIndex={0}
      >
        {controls.length > 0 ? controls : undefined}
        {columns.map(col => this.renderCell(col, rowData))}
      </tr>
    ];

    if (item && item.kind === "success" && isExpanded) {
      row.push(
        <tr key="item-card-viewer">
          <td colSpan={7}>
            <ItemCardViewer item={item.content} />
          </td>
        </tr>
      );
    }

    return row;
  }

  render() {
    const { mapRows } = this.props;
    let content = <div>No Items.</div>;
    if (mapRows) {
      content = (
        <tbody>
          {this.props.mapRows.map((rowData, idx) =>
            this.renderRow(rowData, idx)
          )}
        </tbody>
      );
    }

    return content;
  }
}
