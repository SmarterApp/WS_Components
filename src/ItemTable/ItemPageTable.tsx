import * as React from 'react'
import * as ItemTableHeader from './ItemTableHeader'
import * as ItemTable from './ItemTable'
import * as ItemCardViewModel from '../Models/ItemCardViewModel'
import * as GradeLevels from "../Models/GradeLevels";
import { ItemsSearchViewModel } from "../ScoreGuide/ItemSearchContainer";
import * as ItemModels from '../Models/ItemModels';
import * as AboutItemVM from '../Models/AboutItemVM';
import * as ApiModels from '../Models/ApiModels';

export interface Props {
    onRowSelection: (item: { itemKey: number; bankKey: number }, reset: boolean) => void;
    itemCards?: ItemCardViewModel.ItemCardViewModel[];
    item: ApiModels.Resource<AboutItemVM.AboutThisItem>;
}

export interface State {
    sorts: ItemTableHeader.HeaderSort[];
    selectedRow?: ItemCardViewModel.ItemCardViewModel | null;
}

export class ItemPageTable extends React.Component<Props, State>{
    private headerColumns = ItemTableHeader.headerColumns;
    private dataTableRef: HTMLTableElement;

    constructor(props: Props) {
        super(props);
        this.state = {
            sorts: [],
            selectedRow: null
        }
    }

    onClickHeader = (col: ItemTableHeader.SortColumn) => {
        const newSorts = (this.state.sorts || []).slice();
        const headIdx = newSorts.findIndex(hs => hs.col.header === col.header);
        if (headIdx !== -1) {
            const newSort = Object.assign({}, newSorts[headIdx]);
            if (newSort.direction == ItemTableHeader.SortDirection.Ascending) {
                newSort.direction = ItemTableHeader.SortDirection.Descending;
            }
            else if (newSort.direction == ItemTableHeader.SortDirection.Descending) {
                newSort.direction = ItemTableHeader.SortDirection.NoSort;
            }
            else {
                newSort.direction = ItemTableHeader.SortDirection.Ascending;
            }
            newSorts[headIdx] = newSort;
        } else {
            const newSort: ItemTableHeader.HeaderSort = {
                col: col,
                direction: ItemTableHeader.SortDirection.Ascending,
                resetSortCount: 0
            };
            newSorts.push(newSort);
        }
        this.setState({ sorts: newSorts });
    }

    onSelectItem = (item: ItemCardViewModel.ItemCardViewModel) => {
        const card = { itemKey: item.itemKey, bankKey: item.bankKey }
        if(item === this.state.selectedRow){
            this.props.onRowSelection(card, true)
            this.setState({selectedRow: null})
        }else{
            this.props.onRowSelection(card, false)
            this.setState({selectedRow: item})
        }
    };

    invokeMultiSort(lhs: ItemCardViewModel.ItemCardViewModel, rhs: ItemCardViewModel.ItemCardViewModel): number {
        const sorts = this.state.sorts || [];
        for (const sort of sorts) {
            const diff = sort.col.compare(lhs, rhs) * sort.direction;
            if (diff !== 0) {
                return diff;
            }
        }
        return 0;
    }
    
    getTableData = (): ItemCardViewModel.ItemCardViewModel[] | undefined => {
        let data = this.props.itemCards;
        if (data != undefined) {
            data = this.state.sorts && this.state.sorts.length !== 0
                ? data.sort((lhs, rhs) => this.invokeMultiSort(lhs, rhs))
                : data;
        }
        return data;
    }

    renderTableHeader() {
        return (
            <ItemTableHeader.HeaderTable
                sorts={this.state.sorts}
                onHeaderClick={this.onClickHeader}
                columns={this.headerColumns} />
        );
    }

    renderTable() {
        const itemCards = this.getTableData(); //this returns undefined but in the method it has data. that's w
        let content = (<span className="placeholder-text" role="alert">No results found for the given search terms.</span>);
        if (itemCards != undefined) {
            //if no items are returned we want to return a friendly message
            if (itemCards.length !== 0) {
                content = (
                    <ItemTable.DataTable
                        mapRows={itemCards}
                        rowOnClick={this.onSelectItem}
                        sort={this.state.sorts}
                        tableRef={ref => this.dataTableRef = ref}
                        columns={this.headerColumns}
                        selectedRow={this.state.selectedRow}
                        item={this.props.item}
                    />
                );
            }
        }
        return content;
    }

    render() {
        return (
            <div>
                <table className="item-table mapcomponent-table">
                    {this.renderTableHeader()}
                    {this.renderTable()}
                </table>
            </div>
        );
    }
}
