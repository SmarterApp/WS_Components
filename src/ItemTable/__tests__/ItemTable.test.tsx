import "jsdom-global/register";
import * as React from "react";
import * as ReactDOM from "react-dom";
import * as TestUtils from "react-dom/test-utils";
import { shallow, mount, render } from "enzyme";

import { ItemTable, ItemTableProps } from "../ItemTable";
import { itemCardList } from "../../../mocks/ItemCard/mocks";
import { tabClassNames } from "../../../mocks/ItemTable/mocks";
import { itemHandler } from "./mocks";

import {
  ItemCardModel,
  GradeLevels,
  RubricModel,
  AboutItemModel,
  Resource,
  headerColumns
} from "src/index";

describe("ItemTable", () => {
  const props: ItemTableProps = {
    cardRows: itemCardList,
    onRowExpand: itemHandler,
    onRowSelect: itemHandler,
    sort: [],
    columns: headerColumns,
    isLinkTable: false
  };

  const wrapper = mount(<ItemTable {...props} />);

  it("matches snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("calls onRowSelection()", () => {
    const items = wrapper.find("td.item");
    items.forEach(item => {
      item.simulate("click");
      expect(props.onRowSelect).toHaveBeenCalled();
      expect(props.onRowExpand).toHaveBeenCalled();
    });
  });
});
