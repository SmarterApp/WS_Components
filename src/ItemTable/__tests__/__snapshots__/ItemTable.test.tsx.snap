// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemTable matches snapshot 1`] = `<tbody />`;
